common/dbus/platform/asio_libdbus/dbus_client.cpp

```diff
@@ -0,0 +1,307 @@
+	}
+	if (!dbus_connection_set_timeout_functions(
+			dbus_conn_.get(), AddDBusTimeout, RemoveDBusTimeout, ToggleDBusTimeout, this, NULL)) {
+		dbus_conn_.reset();
+		return MakeError(ConnectionError, "Failed to set timeout functions");
+	}
+	dbus_connection_set_dispatch_status_function(dbus_conn_.get(), HandleDispatch, this, NULL);
+
+	return error::NoError;
+}
+
+void FreeHandlerCopy(void *data) {
+	DBusCallReplyHandler *handler = static_cast<DBusCallReplyHandler *>(data);
+	delete handler;
+}
+
+error::Error DBusClient::CallMethod(
+	const string &destination,
+	const string &path,
+	const string &iface,
+	const string &method,
+	DBusCallReplyHandler handler) {
+	// TODO: with a mutex to make this thread-safe?
+	if (!dbus_conn_) {
+		auto err = InitializeConnection();
+		if (err != error::NoError) {
+			return err;
+		}
+	}
+
+	unique_ptr<DBusMessage, decltype(&dbus_message_unref)> dbus_msg {
+		dbus_message_new_method_call(
+			destination.c_str(), path.c_str(), iface.c_str(), method.c_str()),
+		dbus_message_unref};
+	if (!dbus_msg) {
+		return MakeError(MessageError, "Failed to create new message");
+	}
+
+	DBusPendingCall *pending;
+	if (!dbus_connection_send_with_reply(
+			dbus_conn_.get(), dbus_msg.get(), &pending, DBUS_TIMEOUT_USE_DEFAULT)) {
+		return MakeError(MessageError, "Unable to add message to the queue");
+	}
+
+	// We need to create a copy here because we need to make sure the handler,
+	// which might be a lambda, even with captures, will live long enough for
+	// the finished pending call to use it.
+	DBusCallReplyHandler *handler_copy = new DBusCallReplyHandler(handler);
+	if (!dbus_pending_call_set_notify(pending, HandleReply, handler_copy, FreeHandlerCopy)) {
+		return MakeError(MessageError, "Failed to set reply handler");
+	}
+
+	return error::NoError;
+}
+
+void HandleDispatch(DBusConnection *conn, DBusDispatchStatus status, void *data) {
+	DBusClient *client = static_cast<DBusClient *>(data);
+	if (status == DBUS_DISPATCH_DATA_REMAINS) {
+		// This must give other things in the loop a chance to run because
+		// dbus_connection_dispatch() below can cause this to be called again.
+		client->loop_.Post([conn, status]() {
+			while (dbus_connection_get_dispatch_status(conn) == DBUS_DISPATCH_DATA_REMAINS) {
+				dbus_connection_dispatch(conn);
+			}
+		});
+	}
+}
+
+dbus_bool_t AddDBusWatch(DBusWatch *w, void *data) {
+	// libdbus adds watches in two steps -- using AddDBusWatch() with a disabled
+	// watch which should allocate all the necessary data (and can fail)
+	// followed by ToggleDBusWatch() to enable the watch (see below). We
+	// simplify things for ourselves by ignoring disabled watches and only
+	// actually adding them when ToggleDBusWatch() is called.
+	if (!dbus_watch_get_enabled(w)) {
+		return TRUE;
+	}
+
+	DBusClient *client = static_cast<DBusClient *>(data);
+	asio::posix::stream_descriptor *sd =
+		new asio::posix::stream_descriptor {DBusClient::GetAsioIoContext(client->loop_)};
```

Comment:
Should use `unique_ptr` (evidence: looks like a memory leak at the nearest return below!). If you need to get the ownership back, to pass it to C, you can use `release()`.